### PR TITLE
Client backends for the remote execution of circuits

### PR DESCRIPTION
**Pull request by BrunoLiegiBastonLiegi**
This PR add support for the  `QiboClientBackend` and a `QiskitClientBackend` implemented in [https://github.com/qiboteam/qibo-cloud-backends](https://github.com/qiboteam/qibo-cloud-backends) that allow for the remote execution of circuits. Furthermore, support for QASM `gate` command is added to the parser. 

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi thanks for this, however let me suggest to move these backends directly in https://github.com/qiboteam/qibo-cloud-backends, so we avoid mixing the core code with external providers.

**Comment by BrunoLiegiBastonLiegi** [comment]
Ah ok, sorry about that, I'll move them.

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py:11]: Maybe we should consider using 
```suggestion
def construct_backend(backend, **kwargs):
```

What do you think @scarrazza?

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py:44]: Shouldn't we call it *cloud*, or something like?

`"qibo"` is not much informative (in this context...)

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py:99]: Same proposal of above
```suggestion
    def set_backend(cls, backend, **kwargs):  # pragma: no cover
```

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py:145]: `**kwargs` once more.

File: tests/test_quantum_info_clifford.py
Comment (alecandido) [inline tests/test_quantum_info_clifford.py:113]: Debug prints?

You should drop them before merging.

File: src/qibo/models/circuit.py
Comment (alecandido) [inline src/qibo/models/circuit.py:1252]:
```suggestion
        lines = [line for line in qasm_code.splitlines() if line and line[:2] != "//"]
```
More portable :)

File: src/qibo/models/circuit.py
Comment (alecandido) [inline src/qibo/models/circuit.py:1271]:
```suggestion
            line = line.replace(r"/\s{2,}/g", " ").replace(r"^[\s]+", "")
```

But I really believe we should avoid building a parser procedurally with regexes...

**Review by alecandido — COMMENTED** [review]
Despite the parser being the major upgrade, I didn't pay much attention to it, because we should use a better strategy for defining it (either using the official parser or make our own one, with a specification of a subset of the formal language used).

Using regexes is a simple option for a quick parser, but when things grow and last long it becomes quite involved and complex (and thus much harder to review...).

File: src/qibo/backends/__init__.py
Comment (BrunoLiegiBastonLiegi) [reply src/qibo/backends/__init__.py:44]: I originally called them `qibo-client` and `qiskit-client` but then reverted them to just `qibo` and `qiskit`. Maybe `qibo-cloud` and `qiskit-cloud`?

**Comment by BrunoLiegiBastonLiegi** [comment]
Thanks @alecandido, I agree the parser has to be replaced with something more controllable. If the execution through cloud functionality is not urgent, we could delay this and try to incorporate the use of [https://github.com/openqasm/openqasm](https://github.com/openqasm/openqasm) as we discussed separately.

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi do you prefer to merge this now or wait until qibo-cloud-backens is released? Openqasm could be further upgraded in a separate PR.

**Review by alecandido — APPROVED** [review]

**Comment by BrunoLiegiBastonLiegi** [comment]
I don't know, the only additional thing this PR is providing is the support to the qasm gate command. So it depends, if that is something that is nice to have without waiting for the cloud backends we could merge this.  Otherwise we could wait.

**Comment by scarrazza** [comment]
Ok, lets wait until we test and release the cloud backends.

**Comment by scarrazza** [comment]
Meanwhile, could you please fix tests?

**Comment by BrunoLiegiBastonLiegi** [comment]
Ok I added a reference in the `qibo` api reference docs about `qibo-cloud-backends`, linking to the official documentation (which also contains usage examples). I also added an optional `cloud` group in poetry.
Both this and `qibo-cloud-backends` PR should be ready, we just need to release `qibo-cloud-backends-0.0.1` and then merge this I guess.

File: pyproject.toml
Comment (alecandido) [inline pyproject.toml:101]: Instead of group (that is a Poetry feature, not passing to the package) you should use a pip extra, let's say cloud, such that it will be possible to install with:
```sh
pip install qibo[cloud]
```
https://python-poetry.org/docs/pyproject/#extras

In practice, just move the dependency up, and mark it with `optional = true`. Then, add the extra group, (here it has been never done, and even the other groups are mostly abusing this Poetry feature, i.e. all the cuda groups...).
If you want to have an example you could also have a look to Qibolab and Qibocal.

Comment (alecandido) [reply pyproject.toml]: Solved in 1d12b45

**Comment by alecandido** [comment]
@BrunoLiegiBastonLiegi any known blocker for this?

I believe @MatteoRobbiati could be busy, so I'm adding @stavros11 for the review (and @scarrazza in cc).

File: doc/source/api-reference/qibo.rst
Comment (scarrazza) [inline doc/source/api-reference/qibo.rst:2422]:
```suggestion
Additional backends, that support the remote execution of quantum circuits through cloud service providers, such as IBM and QRC-TII, are provided by the optional qibo plugin `qibo-cloud-backends <https://github.com/qiboteam/qibo-cloud-backends>`_. For more information please refer to the `official documentation <https://qibo.science/qibo-cloud-backends/stable/>`_.
```

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py]: @BrunoLiegiBastonLiegi I'm not sure why you removed this...

Comment (alecandido) [reply src/qibo/backends/__init__.py]: I assumed it was an error due to merging an old version of the `master` branch, and I reintroduced it.

If I was wrong, just let me know.

Comment (BrunoLiegiBastonLiegi) [reply src/qibo/backends/__init__.py]: Mmmh yes I think there was some mess due to the master merge, but I don't think is needed anymore as now we only pass the name of the engine as a string. At least in master it's like that.

Comment (alecandido) [reply src/qibo/backends/__init__.py]: Ok, then I converted `platform` in `kwargs["platform"]`. Maybe it was like that, and somehow changed during the merge.

(I only had a look to the diff, so I did not realize before, sorry)

**Comment by alecandido** [comment]
✅ great to see you again!

**Review by scarrazza — APPROVED** [review]

**Comment by alecandido** [comment]
I rebased on `master`, such that it can be cleanly merged (#1236 was merged in the meanwhile).

Whenever tests are passing, I'd merge this @BrunoLiegiBastonLiegi

**Pull request merged**
